carma-messenger-core/emergency_response_vehicle_plugin/src/emergency_response_vehicle_plugin_node.cpp

```diff
@@ -0,0 +1,457 @@
+    }
+    else{
+      resp->is_successful = false;
+    }
+  }
+
+  void EmergencyResponseVehiclePlugin::arrivedAtEmergencyDestinationCallback(
+    std::shared_ptr<rmw_request_id_t>, 
+    std_srvs::srv::Trigger::Request::SharedPtr req, 
+    std_srvs::srv::Trigger::Response::SharedPtr resp)
+  {
+    RCLCPP_DEBUG_STREAM(rclcpp::get_logger(logger_name_), "ERV has arrived destination. Removing " << route_destination_points_.size() << " route destination points from plugin.");
+
+    // Clear plugin's route member objects
+    route_destination_points_.clear();
+    route_final_destination_name_ = "";
+
+    resp->success = true;
+  }
+
+  void EmergencyResponseVehiclePlugin::incomingEmergencyVehicleResponseCallback(carma_v2x_msgs::msg::EmergencyVehicleResponse::UniquePtr msg)
+  {
+    // Only process message if it is intended for the ego vehicle and the sending vehicle is unable to change lanes
+    if((msg->m_header.recipient_id == bsm_id_string_) && !msg->can_change_lanes){
+
+      // Generate warning message and send to UI so the UI can display an alert for the user
+      carma_msgs::msg::UIInstructions ui_warning_msg;
+
+      ui_warning_msg.msg = "Downstream vehicle " + msg->m_header.sender_id + " is unable to change lanes!";
+      ui_warning_msg.type = carma_msgs::msg::UIInstructions::INFO;
+
+      emergency_vehicle_ui_warnings_pub_->publish(ui_warning_msg);
+
+      // Trigger publication of EmergencyVehicleAck to indicate ERV has received the EmergencyVehicleResponse message
+      broadcastEmergencyVehicleAck(msg->m_header.sender_id);
+    }
+  }
+
+  void EmergencyResponseVehiclePlugin::broadcastEmergencyVehicleAck(const std::string& recipient_id)
+  {
+    carma_v2x_msgs::msg::EmergencyVehicleAck ack_msg;
+
+    ack_msg.m_header.sender_id = bsm_id_string_;
+    ack_msg.m_header.recipient_id = recipient_id;
+    ack_msg.acknowledgement = true;
+
+    outgoing_emergency_vehicle_ack_pub_->publish(ack_msg);
+  }
+
+  double EmergencyResponseVehiclePlugin::getDistanceBetween(const double& lat_1_deg, const double& lon_1_deg, 
+                                    const double& lat_2_deg, const double& lon_2_deg)
+  {
+    // Convert latitude and longitude values from degrees to radians for point 1
+    double lat_1_rad = lat_1_deg * (M_PI / 180.0);
+    double lon_1_rad = lon_1_deg * (M_PI / 180.0);
+
+    // Convert latitude and longitude values from degrees to radians for point 2
+    double lat_2_rad = lat_2_deg * (M_PI / 180.0);
+    double lon_2_rad = lon_2_deg * (M_PI / 180.0);
+
+    // Calculate deltas for inputted latitude and longitude values
+    double delta_lat_rad = lat_2_rad - lat_1_rad;
+    double delta_lon_rad = lon_2_rad - lon_1_rad;
+
+    // Split calculation into two steps for readability
+    double a = std::pow(std::sin(delta_lat_rad / 2.0), 2) + std::pow(std::sin(delta_lon_rad / 2.0), 2) * std::cos(lat_1_rad) * std::cos(lat_2_rad);
+    double distance = EARTH_RADIUS_METERS * 2.0 * std::asin(sqrt(a));
```

Comment:
Is this math correct? I am seeing this on the internet. Or is this the same thing?
```

const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1-a));

const d = R * c; // in metres
```

Reply:
Yes, these evaluate to the same thing. [This answer](https://stackoverflow.com/a/41793716) on StackOverflow is one example that shows how `asin(x)` is equal to `atan2(x, sqrt(1-x*x))`. 

Additionally, a unit test `testGetDistanceBetween` is included to verify the implementation's accuracy with points near TFHRC's campus. The Google Maps measurement tool was used as a baseline and outputted a distance between the points of 334.97 meters, while this Haversine implementation outputted 333.67 meters (99.6% accuracy).

Reply:
okay that's good to know!